File: js/src/resources/text-to-image.ts
import type { HttpClient, PollingOptions, RequestOptions, ActionSchema } from '@runapi.ai/core';
import { compactParams, validateParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import { contract } from '../contract_gen';
import type { TextToImageParams, TextToImageResponse, TaskCreateResponse } from '../types';

const ENDPOINT = '/api/v1/gpt_4o_image/text_to_image';

/**
 * Generates images from a text prompt, optionally guided by source images and a mask.
 * For pure generation, provide a prompt. For editing, provide source_image_urls
 * (and optionally mask_url) to modify specific regions of the source.
 */
export class TextToImage {
  constructor(private readonly http: HttpClient) {}

  /**
   * Generate an image and wait until complete.
   * @param params Text-to-image parameters.
   * @param options Per-request and polling overrides.
   * @returns The completed text-to-image result.
   */
  async run(params: TextToImageParams, options?: RequestOptions & PollingOptions): Promise<TextToImageResponse> {
    const { id } = await this.create(params, options);
    return pollUntilComplete<TextToImageResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
  }

  /**
   * Create a text-to-image task; returns immediately with a task id.
   * @param params Text-to-image parameters.
   * @param options Per-request overrides.
   * @returns The task creation result with id.
   */
  async create(params: TextToImageParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    const body = compactParams(params);
    validateParams(contract['text-to-image'] as ActionSchema, body as Record<string, unknown>);
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body,
      ...options,
    });
  }

  /**
   * Fetch the current status of a text-to-image task.
   * @param id The task id.
   * @param options Per-request overrides.
   * @returns The current text-to-image status.
   */
  async get(id: string, options?: RequestOptions): Promise<TextToImageResponse> {
    return this.http.request<TextToImageResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}

File: ruby/lib/runapi/gpt_4o_image/resources/text_to_image.rb
# frozen_string_literal: true

module RunApi
  module Gpt4oImage
    module Resources
      # GPT-4o Image generation and editing resource.
      # Generate images from text, edit with source images, or inpaint with a mask.
      class TextToImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/gpt_4o_image/text_to_image"

        RESPONSE_CLASS = Types::TextToImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTextToImageResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_contract!(CONTRACT["text-to-image"], params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end
      end
    end
  end
end
